File: py_src/modules/bvh.py
from dataclasses import dataclass
from .hittable import Hittable, HitRecord
from .aabb import aabb, ray_itval, hit_aabb, interval
import time

@dataclass 
class bvh_node(Hittable):
    left: Hittable
    right: Hittable
    bbox: aabb
    
    def __repr__(self) -> str: 
        return f"bvh_node"

def hit_bvh(node: bvh_node, rt: ray_itval, rec: HitRecord) -> bool:
    def hit(node: bvh_node, rt: ray_itval, rec: HitRecord) -> bool:
        from .abs_hit import hit
        if not hit(node.bbox)(rt.r, rt.t): 
            return False
        
        hit_left = hit(node.left)(rt, rec)
        hit_right = hit(node.right)(ray_itval(rt.r, interval(rt.t.lo, rec.t if hit_left else rt.t.hi)), rec)
        
        return (hit_left or hit_right)
    
    t1 = time.time()
    res = hit(node, rt, rec)
    t2 = time.time()
    with open("py_t_real.csv", 'a') as f:
        f.write(f"{t2-t1}\n")
    
    return res

File: cpp_src/src/bvh.cpp
#include "bvh.h"

bool bvh_hit(const bvh_node& node, const ray_itval& rt, HitRecord* rec)  {
    if (!hit_aabb(node.bbox, rt.r, rt.t)) {
        rec->hit = false;
        return false;
    }

    bool hit_left = false;
    bool hit_right = false;

    if (node.left != nullptr) {
        hit_left = node.left.get()->hit(rt, rec);
    }

    if (node.right != nullptr) {
        hit_right = node.right.get()->hit(ray_itval(interval(rt.t.lo, hit_left ? rec->t : rt.t.hi), rt.r), rec);
    }
    
    rec->hit = hit_left || hit_right;
    return rec->hit;
}
